File: apps/website/internal/services/audit.go
package services

import (
	"log"

	"gorm.io/gorm"

	"github.com/tbd-milsim/reforger-backend/internal/models"
)

// WriteAudit appends a row to the audit log. Errors are returned but callers
// generally log-and-continue: an audit failure must not break the primary action.
// A failed write is logged here so the dropped audit trail is at least traceable
// even when the caller ignores the return (T-122 M6).
func WriteAudit(db *gorm.DB, severity models.AuditSeverity, actorID *string, actorName, action, message, targetType, targetID string) error {
	entry := models.AuditLog{
		Severity:   severity,
		ActorID:    actorID,
		ActorName:  actorName,
		Action:     action,
		Message:    message,
		TargetType: targetType,
		TargetID:   targetID,
	}
	if err := db.Create(&entry).Error; err != nil {
		log.Printf("audit write failed: action=%s target=%s/%s: %v", action, targetType, targetID, err)
		return err
	}
	return nil
}

File: apps/website/api_v2/src/services/audit.rs
//! Audit log writer — Rust port of `services/audit.go`.

use sqlx::PgPool;

use crate::models::AuditSeverity;

/// Append a row to the audit log. Best-effort: an audit failure must not break the
/// primary action, so this returns `()` and logs on error (callers ignore it, as in
/// Go). `id` is a bigint sequence; `created_at` is set app-side.
#[allow(clippy::too_many_arguments)]
pub async fn write_audit(
    pool: &PgPool,
    severity: AuditSeverity,
    actor_id: Option<&str>,
    actor_name: &str,
    action: &str,
    message: &str,
    target_type: &str,
    target_id: &str,
) {
    let res = sqlx::query(
        "INSERT INTO audit_logs \
         (severity, actor_id, actor_name, action, message, target_type, target_id, created_at) \
         VALUES ($1, $2, $3, $4, $5, $6, $7, now())",
    )
    .bind(severity)
    .bind(actor_id)
    .bind(actor_name)
    .bind(action)
    .bind(message)
    .bind(target_type)
    .bind(target_id)
    .execute(pool)
    .await;

    if let Err(e) = res {
        tracing::error!(action, error = %e, "audit write failed");
    }
}
